src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameLexer.cs

```diff
@@ -34,39 +30,91 @@ internal Token GetNext()
             return GetNext(out _);
         }
 
+        private static bool IsWhiteSpace(char ch)
+        {
+            switch (ch)
+            {
+                case '\n':
+                case '\r':
+                case ' ':
+                case '\t':
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private char GetNextChar()
+        {
+            char ch;
+            if (_index < _input.Length)
+            {
+                ch = _input[_index++];
+                if (ch == '\0')
+                {
+                    throw new FileLoadException();
```

Comment:
Does the existing unmanaged parser throw a generic `FileLoadException` or is the error message typically customized to talk about invalid assembly name?

I think it would make sense to mention invalid assembly name in the error message.

Reply:
Looks like most of the places use `SR.InvalidAssemblyName`. We may want to do the same here.

Reply:
On 6.0 I get:

```
Unhandled exception. System.IO.FileLoadException: The given assembly name or codebase was invalid. (0x80131047)
   at System.Reflection.AssemblyName.nInit()
   at System.Reflection.AssemblyName..ctor(String assemblyName)
   at Program.<Main>$(String[] args) in C:\Users\vs\source\repos\ConsoleApp6\ConsoleApp6\Program.cs:line 8
```

Reply:
After changes:
```
Unhandled exception. System.IO.FileLoadException: Could not load the specified file.
   at System.Reflection.AssemblyNameLexer.GetNext(String& tokenString)
   at System.Reflection.AssemblyNameParser.Parse(String s)
   at System.Reflection.AssemblyName..ctor(String assemblyName)
   at Program.<Main>$(String[] args) in C:\Users\vs\source\repos\ConsoleApp6\ConsoleApp6\Program.cs:line 8
```

Reply:
The former message is from `FUSION_E_INVALID_NAME`.

Reply:
> Could not load the specified file.

Right, this message is mildly confusing here since somebody may be just trying to parse the assembly, without loading anything.

Reply:
For some reason the CoreRT parser throws `new FileLoadException(SR.InvalidAssemblyName);` from the parser, which matches native behavior 
and `new FileLoadException();` from lexer, which defaults to a different message.

Reply:
There are many places where the exception is thrown from helpers like `private static byte ParseHexNybble(char c)` and the full name is unavailable.

Reply:
Are you suggesting to mention the assembly name only in these two places?

Reply:
Perhaps it would be ok to just fix the 4 cases in lexer to throw `new FileLoadException(SR.InvalidAssemblyName);` - to the match native implementation and to be consistent with the other 17 cases in the parser?

Reply:
To match the current experience, I am suggesting to pass `SR.InvalidAssemblyName` to all these places that throw the FileLoadException in assembly name parsing. Maybe it can be a little helper method that all places call.

To improve the current experience, mentioning the assembly name that failed to parse in the exception message would be nice too. I will leave it up to you whether you want to do it.

Reply:
I'll pass `SR.InvalidAssemblyName` and report the assembly name. I have also merged parser and lexer, so I could further unify error reporting.